api/src/test/java/org/openmrs/api/db/hibernate/HibernateFormDAOTest.java

```diff
@@ -0,0 +1,448 @@
+		containingAllFormFields.add(formService.getFormField(8));
+		Collection<Field> fields = new HashSet<Field>();
+		
+		List<Form> formsActual = hibernateFormDAO.getForms(partialName, published, encounterTypes, retired,
+		    containingAnyFormField, containingAllFormFields, fields);
+		Integer formCountActual = hibernateFormDAO.getFormCount(partialName, published, encounterTypes, retired,
+		    containingAnyFormField, containingAllFormFields, fields);
+		
+		ArrayList<Form> expectedForms = new ArrayList<Form>();
+		
+		Assert.assertEquals(expectedForms, formsActual);
+		Assert.assertEquals(Integer.valueOf(0), formCountActual);
+	}
+	
+	@Test
+	public void shouldReturnZeroFormsWhenMatchingOnNoFields() {
+		FormService formService = Context.getFormService();
+		
+		String partialName = null;
+		Boolean published = null;
+		Collection<EncounterType> encounterTypes = new HashSet<EncounterType>();
+		Boolean retired = null;
+		Collection<FormField> containingAnyFormField = new HashSet<FormField>();
+		Collection<FormField> containingAllFormFields = new HashSet<FormField>();
+		Collection<Field> fields = new HashSet<Field>();
+		fields.add(formService.getField(4));
+		List<Form> formsActual = hibernateFormDAO.getForms(partialName, published, encounterTypes, retired,
+		    containingAnyFormField, containingAllFormFields, fields);
+		Integer formCountActual = hibernateFormDAO.getFormCount(partialName, published, encounterTypes, retired,
+		    containingAnyFormField, containingAllFormFields, fields);
+		
+		ArrayList<Form> expectedForms = new ArrayList<Form>();
+		
+		Assert.assertEquals(expectedForms, formsActual);
+		Assert.assertEquals(Integer.valueOf(0), formCountActual);
+	}
+	
+	@Test
+	public void shouldReturnTwoFormsWhenMatchingOnFields() {
+		FormService formService = Context.getFormService();
+		
+		String partialName = null;
+		Boolean published = null;
+		Collection<EncounterType> encounterTypes = new HashSet<EncounterType>();
+		Boolean retired = null;
+		Collection<FormField> containingAnyFormField = new HashSet<FormField>();
+		Collection<FormField> containingAllFormFields = new HashSet<FormField>();
+		Collection<Field> fields = new HashSet<Field>();
+		fields.add(formService.getField(5));
+		List<Form> formsActual = hibernateFormDAO.getForms(partialName, published, encounterTypes, retired,
+		    containingAnyFormField, containingAllFormFields, fields);
+		Integer formCountActual = hibernateFormDAO.getFormCount(partialName, published, encounterTypes, retired,
+		    containingAnyFormField, containingAllFormFields, fields);
+		
+		ArrayList<Form> expectedForms = new ArrayList<Form>();
+		expectedForms.add(formService.getForm(2));
+		expectedForms.add(formService.getForm(3));
+		Assert.assertEquals(expectedForms, formsActual);
+		Assert.assertEquals(Integer.valueOf(2), formCountActual);
+	}
+	
+	@Test
+	public void shouldReturnTheSameFormMultipleTimesWhenMatchingOnRepeatedFields() {
```

Comment:
I think this shouldn't be the case, it should return a single unique form even if the fields are different as long as they match the same form

Reply:
This test also passes, meaning this is the current implementation of the method. We agree that this is extremely odd, and shouldn't be the case. However, according to the documentation we read about how the getFormCritieria() method and the methods it calls works, this is the correct functionality as written. If this is wrong, then this is a bug/improper implementation of functionality.